### PR TITLE
Sql part2 content

### PR DESCRIPTION
**Pull request by jimflores5**

### CODE REVIEW

File: src/chapters/sql-part2/relationships.rst [hunk 1]
```diff
@@ -0,0 +1,235 @@
+Table Relationships
+===================
+
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst:3]: I like this section, but I found that I needed to read it a few times over to really get what you were getting at. Here are a few suggestions that might help with clarity.

1. Refer to the course offerings as courses. Using "classes" led me to wonder what the difference was and also can lead to confusion since classes have a very specific meaning in programming.
2. An example would be very helpful. If Geometry is in the ``courses`` table, then why do we not need a list of students? Is it because there may be another table for sections of each course that includes that info?

Just food for thought!

File: src/chapters/sql-part2/relationships.rst [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+it makes maintaining that information more difficult.
+
+Any time you add a new column to a table, ask yourself whether the data it
+represents *defines* ``table_name``. If you answer *No*, then store the data in
+a different place (e.g. ``lunch_period`` does not fit within ``students``).
+
+Relating Data
+-------------
+
+Keeping data in separate, smaller chunks is more efficient that keeping a huge
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: "than"

File: src/chapters/sql-part2/relationships.rst [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+amount of data in one place. However, this often means we need to access
+information that exists in different tables.
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: This last sentence needs something to ground it. How about "However, in order to access the information we want, we may have to join tables to get the whole picture."?

File: src/chapters/sql-part2/creating-databases.rst [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+The three tables form a pair of one-to-many relationships. Each
+``writing_supply`` record relates either to several ``pencil_drawer`` entries
+or several ``pen_drawer`` entries.
+
+Each table will need a primary key column, and ``pencil_drawer`` and
+``pen_drawer`` will require foreign key columns. In the query tab of MySQL
+Workbench, ``CREATE`` the tables as described below.
+
+``writing_supply``
+^^^^^^^^^^^^^^^^^^^
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/creating-databases.rst]: Extra carrot!

File: src/chapters/sql-part2/creating-databases.rst [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+#. Clicking *Next* again will import the data. Clicking *Finish* on the final
+   panel returns you to the editor.
+#. In the query tab, run ``SELECT * FROM writing_supply`` to confirm that 6
+   entries now exist in the table.
+#. Repeat the steps 3 - 8 for the ``pencil_drawer`` and ``pen_drawer`` tables.
+
+Confirm that the ``pencil_drawer`` and ``pen_drawer`` tables hold 6 and 15
+entries, respectively. You are now ready to practice more advanced SQL queries.
+
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/creating-databases.rst:123]: Sometimes the CSV import doesn't work properly. Perhaps add a warning with some steps students can take if this process goes wrong?

File: src/chapters/sql-part2/complex-queries.rst [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+included in the result set.
+
+Complex queries often combine joins and subqueries, and they may also include
+multiple ``AND``/``OR`` operators within the ``WHERE`` clause.
+
+Review Joins
+------------
+
+Code along with the following examples. Use the ``writing_supply``,
+``pencil_drawer``, and ``pen_drawer`` tables you created earlier.
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/complex-queries.rst]: Simplify to "Code along with the following examples using the ``storage`` database you created earlier."

**Review by gildedgardenia — CHANGES_REQUESTED** [review]
This looks pretty good! There were very few typos! I found the first page to be confusing and left some suggestions on how to address it.

File: src/chapters/sql-part2/relationships.rst (hunk 1 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst:3]: Good ideas.  I reworked this section a little bit, but I'm not sure if more needs to be done.  Let me know what you think.

File: src/chapters/sql-part2/relationships.rst (hunk 2 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: Fixed.  Thanks.

File: src/chapters/sql-part2/relationships.rst (hunk 3 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: Fixed. Thanks.

File: src/chapters/sql-part2/creating-databases.rst (hunk 4 above)
Comment (jimflores5) [reply src/chapters/sql-part2/creating-databases.rst]: Eaten.  Thanks.

File: src/chapters/sql-part2/complex-queries.rst (hunk 6 above)
Comment (jimflores5) [reply src/chapters/sql-part2/complex-queries.rst]: Done.  Thanks.

File: src/chapters/sql-part2/creating-databases.rst (hunk 5 above)
Comment (jimflores5) [reply src/chapters/sql-part2/creating-databases.rst:123]: I added a troubleshooting section.  Let me know what you think.

File: src/chapters/sql-part2/relationships.rst [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+So, ``courses`` only holds data that directly defines each class the school
+offers. ``teachers`` only contains information that defines each instructor,
+and ``students`` stores student data and nothing else. This sounds obvious, but
+it is very easy to throw lots of data into a table and provide descriptive
+column names. While this may seem like a good idea at the time, it leads to
+repetition of data, and it makes maintaining that information more difficult.
+
+Any time you add a new column to a table, ask yourself whether the data it
+represents *defines* ``table_name``. If you answer *No*, then store the data in
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: "*defines* that table."

File: src/chapters/sql-part2/relationships.rst [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+a different place (e.g. ``lunch_period`` does not fit within ``students``).
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: "``lunch_period`` does not fit within ``students`` so it will have to go in a different table, such as ``schedule``"

File: src/chapters/sql-part2/relationships.rst [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+#. Define a primary key column for each table in the database.
+#. Define a foreign key column for any table that fulfills the *many* role of a
+   one-to-many relationship.
+#. The foreign key of the *many* relates to the primary key of the *one*.
+
+.. admonition:: Tip
+
+   Even if you do not think a table needs one, define a primary key column
+   anyway. As your database and projects grow, you WILL have a need for a unique
+   identifier for each row.
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: I would rephrase this. Reading it again, it sounds off. Maybe instead, "Even if you do not think a table needs one, define a primary key column anyway. Your database needs to grow and be adaptable to change. Adding a primary key column to a table at the beginning helps with that."

File: src/chapters/sql-part2/relationships.rst [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+   Which if the following is the BEST table to store the period and room number
+   for an Algebra I course?
+
+   a. ``teachers``
+   b. ``math_department``
+   c. ``courses``
+   d. ``master_schedule``
+
+.. Answer = d (master_schedule)
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: I would make the answer one of the tables discussed in the reading. Some students might misinterpret what you think the tables should be because the tables are not covered.

File: src/chapters/sql-part2/relationships.rst [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+   For the ``closet``, ``shelf``, and ``box`` tables, what are the
+   relationships? (Don't overthink this).
+
+   a. One-to-many and one-to-many
+   b. One-too-many questions
+   c. Many-too-many-too-many questions
+   d. Um, wait, but... NOOOOOOO! I overthought it!
+
+.. Answer = a (one-to-many and one-to-many)
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/relationships.rst]: I like these questions, but many students would skip over the earlier section covering the topic without a thought. I would either change the questions or change the earlier section to emphasize that students must read the outside articles to answer questions in class.

File: src/chapters/sql-part2/creating-databases.rst [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+   ``writing_supply`` FIRST, before attempting to add data to the other two
+   tables.
+
+.. admonition:: Note
+
+   This is NOT a complete list of everything that could happen to prevent a
+   successful import. Sorry.
+
+   If your issue is not discussed above, YAY! You get to practice your
+   real-world developer skills by figuring out how to make it work!
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/creating-databases.rst]: This note should be reworded to make it more comforting. Maybe simply, "These are ways to address common errors and this is not an exhaustive list of everything that can go wrong with a data import. Don't hesitate to consult the MySQL documentation or your instructor!"

File: src/chapters/sql-part2/complex-queries.rst [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+multiple ``AND``/``OR`` operators within the ``WHERE`` clause.
+
+Review Joins
+------------
+
+Code along with the following examples using the ``storage`` database and
+tables you created earlier.
+
+Inner Join
+^^^^^^^^^^^
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/complex-queries.rst]: Extra carrot!

File: src/chapters/sql-part2/complex-queries.rst [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+
+   The result set contains ``null`` values for any rows that involve pencils. The
+   left join retains all of the data in ``writing_supply``, but it can only
+   combine that information with data from ``pen_drawer`` if the rows share
+   ``supply_id`` values.
+
+   .. figure:: ./figures/left-join-result-set.png
+      :alt: Result set for an inner join with a WHERE clause.
+
+As with inner joins, we can restrict the size of the result set:
```
Comment (gildedgardenia) [inline src/chapters/sql-part2/complex-queries.rst]: I would specify that the size of the result set is decreasing because of the additional condition.

**Review by gildedgardenia — APPROVED** [review]
I went over some things that could be confusing to people, but otherwise approved!

File: src/chapters/sql-part2/relationships.rst (hunk 7 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: Fixed.  Thanks.

File: src/chapters/sql-part2/relationships.rst (hunk 8 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: Done.  Thanks.

File: src/chapters/sql-part2/relationships.rst (hunk 9 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: Changes applied.  Thanks.

File: src/chapters/sql-part2/relationships.rst (hunk 10 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: Fixed.  Thanks.

File: src/chapters/sql-part2/relationships.rst (hunk 11 above)
Comment (jimflores5) [reply src/chapters/sql-part2/relationships.rst]: I reworded the previous section to emphasize reading about one-to-one and many-to-many relationships.

File: src/chapters/sql-part2/creating-databases.rst (hunk 12 above)
Comment (jimflores5) [reply src/chapters/sql-part2/creating-databases.rst]: Changes applied.  Thanks.

File: src/chapters/sql-part2/complex-queries.rst (hunk 13 above)
Comment (jimflores5) [reply src/chapters/sql-part2/complex-queries.rst]: Surplus veggie consumed.

File: src/chapters/sql-part2/complex-queries.rst (hunk 14 above)
Comment (jimflores5) [reply src/chapters/sql-part2/complex-queries.rst]: Done.  Thanks.
